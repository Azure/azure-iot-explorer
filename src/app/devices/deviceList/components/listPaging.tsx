/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { ActionButton } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import '../../../css/_deviceListPaging.scss';

export interface ListPagingDataProps {
    continuationTokens: string[];
    currentPageIndex: number;
}
export interface ListPagingDispatchProps {
    fetchPage: (pageNumber: number) => void;
}
export const ListPaging: React.FC<ListPagingDataProps & ListPagingDispatchProps> = (props: ListPagingDataProps & ListPagingDispatchProps) => {
    const { t } = useTranslation();
    const { continuationTokens, currentPageIndex, fetchPage } = props;

    const renderSection = () => {
        return (
            <section role="navigation" className="grid-paging">
                <span className="pages">{t(ResourceKeys.deviceLists.paging.pages)}</span>
                <div role="list" className="page-list">
                    {continuationTokens.map(renderListItem)}
                </div>
            </section>
        );
    };

    const renderListItem = (continuationToken: string, index: number) => {
        const onFetchPageClick = () => {
            fetchPage(index);
        };

        return (
            <div
                key={`page_${index}`}
                role="listitem"
                className={index === currentPageIndex ? 'selected' : ''}
            >
                {index !== currentPageIndex ?
                    (<ActionButton
                        onClick={onFetchPageClick}
                    >
                        {index === continuationTokens.length - 1 ? 'Â»' : index + 1}
                    </ActionButton>) :
                    (<span>{index + 1}</span>)
                }
            </div>
        );
    };

    return (
            (continuationTokens && continuationTokens.length > 0 && renderSection()) || <></>
        );
};
