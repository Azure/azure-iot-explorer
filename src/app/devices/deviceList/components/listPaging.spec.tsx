/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { ActionButton } from '@fluentui/react';
import { ListPaging } from './listPaging';

describe('components/devices/listPaging', () => {

    it('renders nothing with empty continuationTokens', () => {
        const component = (
            <ListPaging
                continuationTokens={null}
                currentPageIndex={0}
                fetchPage={jest.fn()}
            />
        );
        expect(shallow(component)).toMatchSnapshot();
    });

    it('matches snapshot with continuation tokens', () => {
        expect(shallow(
            <ListPaging
                continuationTokens={['', 'abc123']}
                currentPageIndex={0}
                fetchPage={jest.fn()}
            />
        )).toMatchSnapshot();
    });

    it('matches snapshot on different page', () => {
        expect(shallow(
            <ListPaging
                continuationTokens={['', 'abc123', 'def456']}
                currentPageIndex={1}
                fetchPage={jest.fn()}
            />
        )).toMatchSnapshot();
    });

    it('calls fetchPage on click', () => {
        const fetchPage = jest.fn();
        const wrapper = mount(
        <ListPaging
            continuationTokens={['', 'abc123', 'def456']}
            currentPageIndex={1}
            fetchPage={fetchPage}
        />);

        const link = wrapper.find(ActionButton).first();
        link.simulate('click');
        expect(fetchPage).toHaveBeenCalled();
    });
});
