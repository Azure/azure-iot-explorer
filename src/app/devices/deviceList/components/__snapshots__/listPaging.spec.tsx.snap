// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/listPaging matches snapshot on different page 1`] = `
<section
  className="grid-paging"
  role="navigation"
>
  <span
    className="pages"
  >
    deviceLists.paging.pages
  </span>
  <div
    className="page-list"
    role="list"
  >
    <div
      className=""
      key="page_0"
      role="listitem"
    >
      <CustomizedActionButton
        onClick={[Function]}
      >
        1
      </CustomizedActionButton>
    </div>
    <div
      className="selected"
      key="page_1"
      role="listitem"
    >
      <span>
        2
      </span>
    </div>
    <div
      className=""
      key="page_2"
      role="listitem"
    >
      <CustomizedActionButton
        onClick={[Function]}
      >
        »
      </CustomizedActionButton>
    </div>
  </div>
</section>
`;

exports[`components/devices/listPaging matches snapshot with continuation tokens 1`] = `
<section
  className="grid-paging"
  role="navigation"
>
  <span
    className="pages"
  >
    deviceLists.paging.pages
  </span>
  <div
    className="page-list"
    role="list"
  >
    <div
      className="selected"
      key="page_0"
      role="listitem"
    >
      <span>
        1
      </span>
    </div>
    <div
      className=""
      key="page_1"
      role="listitem"
    >
      <CustomizedActionButton
        onClick={[Function]}
      >
        »
      </CustomizedActionButton>
    </div>
  </div>
</section>
`;

exports[`components/devices/listPaging renders nothing with empty continuationTokens 1`] = `<Fragment />`;
